File: packages/engine/src/core/index-tail.ts
import { closeSync, fstatSync, openSync, readSync } from 'node:fs'
import { validatePayload } from '@sofar/schema'
import { validateEnvelope } from './envelope'
import { cursorUsable, logStat, logUntouched, type InitiativeCursor } from './index-store'

/**
 * Read only what a log has grown by (record-index 1.2).
 *
 * This is where O(new events) actually happens. Every other cross-record
 * surface in the codebase re-reads whole logs to answer questions about a
 * handful of recent events; with a cursor, the common case — nothing appended
 * since last time — costs one stat and reads nothing at all.
 *
 * The correctness bargain is that a seek is never TRUSTED. The cursor stores
 * the byte offset where its event's line begins, so the first line read back
 * must carry that exact id. When it does, the offset described this file and
 * the tail after it is genuinely new. When it does not — a rewritten log, an
 * import, a restore, a truncation that happened to land plausibly — the reader
 * silently falls back to reading everything, which is slower and right. There
 * is no configuration for this and no way to force the fast path: an index
 * that can be talked into a wrong answer is worse than no index.
 */

/** A decoded envelope, only as far as the index needs it. */
export interface IndexedEvent {
  id: string
  type: string
  session: string
  initiative: string
  payload: Record<string, unknown>
  ts: string
}

export interface TailRead {
  events: IndexedEvent[]
  /** Cursor to persist, or null when the log held no usable event. */
  cursor: InitiativeCursor | null
  /** True when the whole log was read — a cold start or a failed corroboration. */
  full: boolean
}

/**
 * A line the index may anchor a cursor on, and whether the fold would REPLAY
 * it.
 *
 * The two are not the same question, and the fold answers them separately:
 * a line that fails envelope validation is not an event at all, while a
 * payload-invalid or unknown-typed event is a real event the replay skips.
 * The cursor tracks the former (it is a position in a file) and `usable`
 * marks the latter, so a derivation built from these events sees exactly the
 * set the fold sees — no more, which would invent work, and no fewer, which
 * would hide it.
 */
function decode(line: string): { event: IndexedEvent; usable: boolean } | null {
  let raw: unknown
  try {
    raw = JSON.parse(line)
  } catch {
    return null // a corrupt line is skipped, exactly as the fold skips it
  }
  const check = validateEnvelope(raw)
  if (!check.ok) return null

  const e = check.event
  return {
    event: {
      id: e.id,
      type: e.type,
      session: e.session,
      initiative: e.initiative,
      payload: e.payload,
      ts: e.ts,
    },
    usable: validatePayload(e.type, e.payload).ok,
  }
}

/** Read `length` bytes from `start`, or the whole file when start is 0. */
function readFrom(path: string, start: number): { text: string; size: number; mtimeMs: number } | null {
  let fd: number | null = null
  try {
    fd = openSync(path, 'r')
    // One fstat for both, so the stored cursor describes exactly the bytes
    // this call read — a separate stat could straddle a concurrent append.
    const stat = fstatSync(fd)
    const size = stat.size
    const mtimeMs = stat.mtimeMs
    if (start >= size) return { text: '', size, mtimeMs }
    const length = size - start
    const buf = Buffer.allocUnsafe(length)
    readSync(fd, buf, 0, length, start)
    return { text: buf.toString('utf8'), size, mtimeMs }
  } catch {
    return null
  } finally {
    if (fd !== null) {
      try {
        closeSync(fd)
      } catch {
        // Cannot change the bytes already read.
      }
    }
  }
}

/**
 * Split into lines, keeping each line's absolute byte offset.
 *
 * Offsets are computed in BYTES, not characters. Event payloads carry prose,
 * and prose carries em dashes and arrows — a UTF-16 length would drift from
 * the file position on the first non-ASCII character and put every subsequent
 * cursor a few bytes off.
 */
function linesWithOffsets(text: string, base: number): { line: string; offset: number }[] {
  const out: { line: string; offset: number }[] = []
  let cursor = base
  for (const line of text.split('\n')) {
    if (line.trim().length > 0) out.push({ line, offset: cursor })
    cursor += Buffer.byteLength(line, 'utf8') + 1 // +1 for the newline
  }
  return out
}

/** Events appended since the cursor, plus the cursor to store next time. */
export function readSince(logPath: string, cursor: InitiativeCursor | null): TailRead {
  const stat = cursor === null ? null : logStat(logPath)

  // Nothing was appended and nothing was rewritten — the cheapest answer
  // there is, and the one a quiet initiative gives on every single refresh.
  // A record's logs are overwhelmingly quiet at any instant, so this is the
  // common case, not the optimization: it is what keeps a refresh O(active
  // initiatives) instead of O(initiatives).
  if (cursor !== null && stat !== null && logUntouched(stat, cursor)) {
    return { events: [], cursor, full: false }
  }

  if (cursor !== null && stat !== null && cursorUsable(stat, cursor)) {
    const chunk = readFrom(logPath, cursor.offset)
    if (chunk !== null) {
      const lines = linesWithOffsets(chunk.text, cursor.offset)
      const first = lines[0] === undefined ? null : decode(lines[0].line)
      // Corroboration: the offset must land exactly on the event it claims.
      if (first !== null && first.event.id === cursor.id) {
        const fresh = lines.slice(1)
        const events: IndexedEvent[] = []
        let last = { id: cursor.id, offset: cursor.offset }
        for (const { line, offset } of fresh) {
          const decoded = decode(line)
          if (decoded === null) continue
          last = { id: decoded.event.id, offset }
          if (decoded.usable) events.push(decoded.event)
        }
        return { events, cursor: { ...last, size: chunk.size, mtimeMs: chunk.mtimeMs }, full: false }
      }
    }
  }

  // Cold start, or the offset did not describe this file. Read everything.
  const whole = readFrom(logPath, 0)
  if (whole === null) return { events: [], cursor: null, full: true }
  const events: IndexedEvent[] = []
  let last: { id: string; offset: number } | null = null
  for (const { line, offset } of linesWithOffsets(whole.text, 0)) {
    const decoded = decode(line)
    if (decoded === null) continue
    last = { id: decoded.event.id, offset }
    if (decoded.usable) events.push(decoded.event)
  }
  return {
    events,
    cursor: last === null ? null : { ...last, size: whole.size, mtimeMs: whole.mtimeMs },
    full: true,
  }
}

File: crates/sofar-core/src/index_tail.rs
//! Read only what a log has grown by (`core/index-tail.ts`, record-index
//! 1.2): a cursor's offset is corroborated by the id at that line, never
//! trusted; anything else is a full read.

use std::fs::File;
use std::io::{Read as _, Seek as _, SeekFrom};
use std::path::Path;

use crate::envelope::validate_envelope;
use crate::index_store::{Cursor, LogStat, cursor_usable, log_stat, log_untouched, mtime_ms_of};
use crate::json::{self, Json, Object};
use crate::payload::validate_payload;
use crate::text::js_trim;

/// A decoded envelope, only as far as the index needs it.
#[derive(Debug, Clone, PartialEq)]
pub struct IndexedEvent {
    pub id: String,
    pub event_type: String,
    pub session: String,
    pub initiative: String,
    pub payload: Object,
    pub ts: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct TailRead {
    pub events: Vec<IndexedEvent>,
    /// Cursor to persist, or None when the log held no usable event.
    pub cursor: Option<Cursor>,
    /// True when the whole log was read.
    pub full: bool,
}

/// `decode`: an envelope-valid line and whether the fold would replay it.
fn decode(line: &str) -> Option<(IndexedEvent, bool)> {
    let raw = json::parse(line).ok()?;
    let e = validate_envelope(raw).ok()?;
    let usable = validate_payload(&e.event_type, &Json::Obj(e.payload.clone())).is_ok();
    Some((
        IndexedEvent {
            id: e.id,
            event_type: e.event_type,
            session: e.session,
            initiative: e.initiative,
            payload: e.payload,
            ts: e.ts,
        },
        usable,
    ))
}

struct Chunk {
    text: String,
    stat: LogStat,
}

/// `readFrom`: bytes from `from` to the end, with the stat of the same open file.
fn read_from(path: &Path, from: u64) -> Option<Chunk> {
    let mut file = File::open(path).ok()?;
    let m = file.metadata().ok()?;
    let stat = LogStat {
        size: m.len(),
        mtime_ms: mtime_ms_of(&m),
    };
    if from >= stat.size {
        return Some(Chunk {
            text: String::new(),
            stat,
        });
    }
    file.seek(SeekFrom::Start(from)).ok()?;
    let mut buf = Vec::with_capacity(usize::try_from(stat.size - from).ok()?);
    file.read_to_end(&mut buf).ok()?;
    Some(Chunk {
        text: String::from_utf8_lossy(&buf).into_owned(),
        stat,
    })
}

/// `linesWithOffsets`: non-blank lines with their absolute BYTE offsets.
fn lines_with_offsets(text: &str, base: u64) -> Vec<(&str, u64)> {
    let mut out = Vec::new();
    let mut cursor = base;
    for line in text.split('\n') {
        if !js_trim(line).is_empty() {
            out.push((line, cursor));
        }
        cursor += line.len() as u64 + 1;
    }
    out
}

/// `readSince`: events appended since the cursor, plus the cursor to store next.
#[must_use]
pub fn read_since(log_path: &Path, cursor: Option<&Cursor>) -> TailRead {
    let stat = cursor.and_then(|_| log_stat(log_path));
    if let (Some(cursor), Some(stat)) = (cursor, stat) {
        if log_untouched(stat, cursor) {
            return TailRead {
                events: Vec::new(),
                cursor: Some(cursor.clone()),
                full: false,
            };
        }
        if cursor_usable(stat, cursor)
            && let Some(chunk) = read_from(log_path, cursor.offset)
        {
            let lines = lines_with_offsets(&chunk.text, cursor.offset);
            let first = lines.first().and_then(|(line, _)| decode(line));
            if let Some((first, _)) = first
                && first.id == cursor.id
            {
                let mut events = Vec::new();
                let mut last = (cursor.id.clone(), cursor.offset);
                for (line, offset) in &lines[1..] {
                    let Some((event, usable)) = decode(line) else {
                        continue;
                    };
                    last = (event.id.clone(), *offset);
                    if usable {
                        events.push(event);
                    }
                }
                return TailRead {
                    events,
                    cursor: Some(Cursor {
                        id: last.0,
                        offset: last.1,
                        size: chunk.stat.size,
                        mtime_ms: chunk.stat.mtime_ms,
                        max_id: None,
                        voided: None,
                    }),
                    full: false,
                };
            }
        }
    }
    let Some(whole) = read_from(log_path, 0) else {
        return TailRead {
            events: Vec::new(),
            cursor: None,
            full: true,
        };
    };
    let mut events = Vec::new();
    let mut last: Option<(String, u64)> = None;
    for (line, offset) in lines_with_offsets(&whole.text, 0) {
        let Some((event, usable)) = decode(line) else {
            continue;
        };
        last = Some((event.id.clone(), offset));
        if usable {
            events.push(event);
        }
    }
    TailRead {
        events,
        cursor: last.map(|(id, offset)| Cursor {
            id,
            offset,
            size: whole.stat.size,
            mtime_ms: whole.stat.mtime_ms,
            max_id: None,
            voided: None,
        }),
        full: true,
    }
}
